db/db_impl_write.cc

```diff
@@ -392,15 +415,31 @@ Status DBImpl::WriteToWAL(const autovector<WriteThread::Writer*>& write_group,
     }
   }
 
+  if (concurrent) {
+    // We need to lock log_write_mutex_ since logs_ and alive_log_files might be
+    // pushed back concurrently
+    log_write_mutex_.Lock();
+  }
+  *last_sequence = versions_->FetchAddLastToBeWrittenSequence(total_count);
+  auto sequence = *last_sequence + 1;
   WriteBatchInternal::SetSequence(merged_batch, sequence);
 
   Slice log_entry = WriteBatchInternal::Contents(merged_batch);
+  log::Writer* log_writer = logs_.back().writer;
   status = log_writer->AddRecord(log_entry);
+  if (log_used != nullptr) {
+    *log_used = logfile_number_;
+  }
   total_log_size_ += log_entry.size();
   alive_log_files_.back().AddSize(log_entry.size());
+  if (concurrent) {
```

Comment:
I suggest we put line 423 to 434 to a separate helper function, and then directly call it from a function like DBImpl::PrepareToWALBuffer() where the mutex is protected. And we can do the same for line 470-478.